### ISSUE TITLE
Unable to configure dump plugin with git source

### ISSUE THREAD

**Issue by mthaddon** [canonical/charmcraft#1103]
I've created https://github.com/mthaddon/charmcraft-test to show the problem here. I'm experiencing with adding a git source to my `charmcraft.yaml`. The code is downloaded, but charmcraft errors, and the charm refuses to build.

Can anyone help me figure out if this is a configuration issue or a bug in charmcraft?

**Comment by sergiusens** [comment]
Looks more like an issue with the source handles, to prove it, I changed the plugin to `nil`

```diff
diff --git a/charmcraft.yaml b/charmcraft.yaml
index 8bb7a4b..058543e 100644
--- a/charmcraft.yaml
+++ b/charmcraft.yaml
@@ -15,6 +15,6 @@ parts:
     source-depth: 1
     source-branch: master
     source-type: git
-    plugin: dump
+    plugin: nil
     organize:
       '*': lib/juju-upgrader/
```

Here's the output
```
sergiusens@imladris:~/repo/charms/charmcraft-test$ charmcraft pack
Packing the charm                                                                                   
Parts processing error: Failed to pull source: command ['git', 'clone', '--recursive', '--branch', 'master', '--depth', '1', 'https://git.launchpad.net/juju-upgrader', '/root/parts/git-stuff/src'] exited with code 1.
Make sure sources are correctly specified.                                           
Failed to build charm for bases index '0'.                                                          
Full execution log: '/home/sergiusens/.local/state/charmcraft/log/charmcraft-20230512-084624.069424.log'                                                                                                 
```

The interesting bits from the logs are:

```
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:39.813 :: Cloning into '/root/parts/git-stuff/src'...
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: usage: git submodule [--quiet] [--cached]
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] add [-b <branch>] [-f|--force] [--name <name>] [--reference <repository>] [--] <repository> [<path>]
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] status [--cached] [--recursive] [--] [<path>...]
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] init [--] [<path>...]
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] deinit [-f|--force] (--all| [--] <path>...)
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] update [--init] [--remote] [-N|--no-fetch] [-f|--force] [--checkout|--merge|--rebase] [--[no-]recommend-shallow] [--reference <repository>] [--recursive] [--] [<path>...]
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] set-branch (--default|--branch <branch>) [--] <path>
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] set-url [--] <path> <newurl>
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] summary [--cached|--files] [--summary-limit <n>] [commit] [--] [<path>...]
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] foreach [--recursive] <command>
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] sync [--recursive] [--] [<path>...]
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.921 :: or: git submodule [--quiet] absorbgitdirs [--] [<path>...]
2023-05-12 08:46:49.402 :: 2023-05-12 11:46:48.922 Parts processing error: Failed to pull source: command ['git', 'clone', '--recursive', '--branch', 'master', '--depth', '1', 'https://git.launchpad.net/juju-upgrader', '/root/parts/git-stuff/src'] exited with code 1.
```

I may be way off here as I am not sure how `clone --recursive` works internally, but it seem to be calling `git submodule` incorrectly.

The code would exist as the first clone is there. The default behavior is to try and recurse clone all submodules if `source-submodules` is not set. This can be overridden with:

```diff
diff --git a/charmcraft.yaml b/charmcraft.yaml
index 8bb7a4b..cb103ba 100644
--- a/charmcraft.yaml
+++ b/charmcraft.yaml
@@ -15,6 +15,7 @@ parts:
     source-depth: 1
     source-branch: master
     source-type: git
+    source-submodules: []
     plugin: dump
     organize:
       '*': lib/juju-upgrader/
```

Which leads me to a new error in the test project, but unrelated (as only one part exists)

```
- metadata: Cannot read the metadata.yaml file. (https://juju.is/docs/sdk/charmcraft-analyzers-and-linters#heading--metadata)                                                                             
```

**Comment by mthaddon** [comment]
Awesome, thx. I can confirm I can build the charm I originally found this problem in by adding `source-submodules: []`.

**Comment by lengau** [comment]
This appears to be a bug in the included version of git (from core20) that doesn't occur with the most recent git. The likely resolution to this is going to come when we migrate charmcraft to core22.

**Comment by lengau** [comment]
@mthaddon this should be fixed in edge and in the upcoming 2.5 release. Please let us know if it's not.
